client/requirements.txt

```diff
@@ -1,2 +1,3 @@
+pyjwt
```

Comment:
Line no. 4 of `lib/pbench/test/functional/server/test_connect.py` imports OpenIDClient of the auth module "`from pbench.server.auth import OpenIDClient`" -> Line no. 6 of `lib/pbench/server/auth/__init__.py` imports jwt.

Reply:
Why is this being added now with the import of `jwt` at line 7 of `lib/pbench/server/auth/__init__.py` existed before this commit?

Reply:
Because our server functional tests never directly used `auth` module before. Functional tests use `lib/pbench/client` and if you look under the `pbench.client` module you would see that we didn't use the auth module anywhere so there was no need before this PR to add `pyjwt` to the `client/requirements.txt` file.

Reply:
So, a functional _test_ won't run because it pulls in _Server_ code without having the **Server** requirements met...that sounds highly confused to me.

A proper functional test should only be accessing the Server via the RESTful API (I think...), so it shouldn't need to import any Server modules (and therefore shouldn't need to satisfy the _Server's_ requirements).  What is the Pbench Client trying to do which no other client (e.g., one without access to the Pbench Server code) would need or want to do??

If a test needs to reference Server modules, then that sounds like a _unit_ test to me.  (This will be less true once we export Python APIs for the Agent (at which point we'll have to understand how to meet the code's requirements), but I think it's accurate enough for now.)

Now, if there is some bit of functionality in `lib/pbench/server/auth/__init__.py` which is of general use (or, particularly, is of good use to the Pbench Client), then perhaps it should be split off into a "common" code module (and, we'll need to figure out how to satisfy its requirements).  But, really, the Pbench Client shouldn't be "special", so whatever it needs is likely to be needed by any other client, and we need to understand how to satisfy that need!

Reply:
> So, a functional _test_ won't run because it pulls in _Server_ code without having the **Server** requirements met...that sounds highly confused to me.

Well, at the moment we can not perform the functional test on the OIDC broker (Keycloak server) by hitting any APIs and until we have appropriate APIs for doing that, we have to pull the server code related to OIDC broker that we want to test.
Also, I am not sure how much API access we will be giving for the Keycloak broker via pbench-server so that's something we have to decide and how we want to implement it. Dashboard will likely be present somewhere in this logical flow to generate tokens and authorization but we don't have the mechanism to do that in our functional test yet. 

> Now, if there is some bit of functionality in `lib/pbench/server/auth/__init__.py` which is of general use (or, particularly, is of good use to the Pbench Client), then perhaps it should be split off into a "common" code module (and, we'll need to figure out how to satisfy its requirements). But, really, the Pbench Client shouldn't be "special", so whatever it needs is likely to be needed by any other client, and we need to understand how to satisfy that need!

I agree, I think the `OIDCClient` related code can be refactored into a `common` directory. I will create a Jira issue to work on it.

Reply:
I think the problem is that the test in question (`lib/pbench/test/functional/server/test_connect.py:TestConnect.test_keycloak()`) is not actually testing the Pbench Server -- rather, it's trying to test a bit of the Pbench Server code in isolation -- and that doesn't really fit into the model that we're building.

That's really more like a unit test than a functional test.  So, unless the `OIDCCLient` object is something that we expect other clients of the Pbench Server to need, I would not refactor it into a `common` directory.

I think the real issue here is just that we're getting ahead of ourselves.  We should have **_unit_** tests for the `OIDCClient` code, and then we should have functional tests for the Pbench Server.  Between the two, that should provide adequate coverage of our code.

I can understand how, for development purposes, you might want to have a little Python program which exercises `OIDCClient` against a real Keycloak server, but I don't think we'll need that in the long term -- that should be subsumed into the Pbench Server code, and the Pbench Server functional tests should exercise it sufficiently to ensure that it operates correctly.

Reply:
> I think the real issue here is just that we're getting ahead of ourselves. We should have **_unit_** tests for the `OIDCClient` code, and then we should have functional tests for the Pbench Server. Between the two, that should provide adequate coverage of our code.

Sure thing. But I also want to make sure the Keycloak container built from our pod is accepting the connections and is healthy and that is only possible with functional tests (because there is no entity outside the pbench-server that is using this container at the moment). Until we have the dashboard in our functional test to use the OIDC broker we might want to do some isolated functional tests on the broker to make sure everything is as expected isn't it? Also, are you suggesting that we should get rid of `test_keycloak` function from our functional test?

> I can understand how, for development purposes, you might want to have a little Python program which exercises `OIDCClient` against a real Keycloak server, but I don't think we'll need that in the long term -- that should be subsumed into the Pbench Server code, and the Pbench Server functional tests should exercise it sufficiently to ensure that it operates correctly.

We wouldn't need that for the long run, yes, but right now we have no mechanism in place to test the real authentication server. When we have an actual end-to-end functional test then of course we shouldn't do the isolated functional test on the Keycloak server because that will be tested under the hood anyway.

Reply:
I don't think that our test system should be testing the Keycloak server.  (That's an issue for our DevOps, not for our code development pipeline.)  Once we're in steady state, it should be reasonable for us to assume that the Keycloak server is working, and so what we want to test is whether the Pbench Server behaves properly when it uses it.

So, I don't think we (should) need anything in our (Pbench Server) tests which pokes the Keycloak server directly (at least, we shouldn't need anything that any other Pbench Server client would not do).  Yes, in the interim, we'll want something to satisfy us (er, you) that the Keycloak deployment is working, but that doesn't (or shouldn't) need to be part of the test system.  That's just something we'll want until we gain confidence in our ability to deploy and use Keycloak, but then we won't need it anymore.

As you say, once we have the appropriate functionality in the Dashboard and appropriate functional tests for it in the CI, then we'll be exercising Keycloak that way.  But, in the interim, I would just skip that -- the Pbench Server doesn't (yet) have any functionality which depends on Keycloak, so we don't need any functional tests which involve it.  (So, yeah, I would drop the `test_keycloak()` test; we have unit tests for `OIDCClient`, right?  Those should be all we need until the Pbench Server provides functionality which uses it.)

Reply:
Alright, I think you convinced me to remove it at least from this PR. But, I guess if we ever want to do some isolated e-2-e tests on the Keycloak server we will create a subdirectory under the `functional` and add tests there, so they will be easier to remove when we have the components that use the OIDC functionality in our functional tests.